src/codegen/dyngen.rs

```diff
@@ -0,0 +1,181 @@
+
+    /// Tracks the information that is passed to the library struct at the end of the `::new`
+    /// constructor, e.g.:
+    /// ```ignore
+    /// impl LibFoo {
+    ///     pub unsafe fn new<P>(path: P) -> Result<Self, ::libloading::Error>
+    ///     where
+    ///         P: AsRef<::std::ffi::OsStr>,
+    ///     {
+    ///         ...
+    ///         Ok(LibFoo {
+    ///             __library: __library,
+    ///             foo,
+    ///             bar, // <- tracks these
+    ///             ...
+    ///         })
+    ///     }
+    /// }
+    /// ```
+    init_fields: Vec<proc_macro2::TokenStream>,
+}
+
+impl Default for DynamicItems {
+    fn default() -> Self {
+        DynamicItems {
+            struct_members: vec![],
+            struct_implementation: vec![],
+            runtime_checks: vec![],
+            constructor_inits: vec![],
+            init_fields: vec![],
+        }
+    }
+}
+
+impl DynamicItems {
+    pub fn new() -> Self {
+        Self::default()
+    }
+
+    pub fn get_tokens(
+        &self,
+        lib_ident: Ident,
+        check_struct_ident: Ident,
+    ) -> proc_macro2::TokenStream {
+        let struct_members = &self.struct_members;
+        let constructor_inits = &self.constructor_inits;
+        let init_fields = &self.init_fields;
+        let struct_implementation = &self.struct_implementation;
+        let runtime_checks = &self.runtime_checks;
+        quote! {
+            extern crate libloading;
+
+            pub struct #lib_ident {
+                __library: ::libloading::Library,
+                #(#struct_members)*
+            }
+
+            impl #lib_ident {
+                pub unsafe fn new<P>(
+                    path: P
+                ) -> Result<Self, ::libloading::Error>
+                where P: AsRef<::std::ffi::OsStr> {
+                    let __library = ::libloading::Library::new(path)?;
+                    #( #constructor_inits )*
+                    Ok(
+                        #lib_ident {
+                            __library: __library,
+                            #( #init_fields ),*
+                        }
+                    )
+                }
+
+                pub fn can_call(&self) -> #check_struct_ident {
+                    #check_struct_ident { __library: self }
+                }
+
+                #( #struct_implementation )*
+            }
+
+            pub struct #check_struct_ident<'a> {
+                __library: &'a #lib_ident,
+            }
+
+            impl<'a> #check_struct_ident<'a> {
+                #( #runtime_checks )*
+            }
+        }
+    }
+
+    pub fn add_function(
+        &mut self,
+        ident: Ident,
+        abi: Abi,
+        args: Vec<proc_macro2::TokenStream>,
+        args_identifiers: Vec<proc_macro2::TokenStream>,
+        ret: proc_macro2::TokenStream,
+        ret_ty: proc_macro2::TokenStream,
+    ) {
+        assert_eq!(args.len(), args_identifiers.len());
```

Comment:
Just a note, this seems to assert on varargs parameters.

Not sure if those are not supported widely and this is expected? At least things *seem* to work when I remove the assert. The following snippet will trigger the assert.

```
int thing(int, ...);
```